R/estimate_uncertainty.R

```diff
@@ -0,0 +1,213 @@
+#'   estimates an independent negative binomial dispersion parameter. This code
+#'   was adapted from code written (under an MIT license) by the Karlsruhe
+#'   Institute of Technology RESPINOW German Hospitalization Nowcasting Hub,
+#' Modified from: https://github.com/KITmetricslab/RESPINOW-Hub/blob/39e2b17bc79492b0aee4c0b615a1c8dbf978ef53/code/baseline/functions.R#L142 #nolint
+#' @param triangle_to_nowcast matrix of the incomplete reporting triangle to be
+#'   nowcasted, with rows representing the time points of reference and columns
+#'   representing the delays
+#' @param delay_pmf vector of delays assumed to be indexed starting at the
+#'   first delay column in `triangle_to_nowcast`
+#' @param n_history_dispersion integer indicating the number of reference dates
+#'   to be used in the estimate of the dispersion, always starting from the most
+#'   recent refrence date. The default is to use the whole reporting triangle,
+#'   so `nrow(triangle_to_nowcast) - 1`
+#' @importFrom cli cli_abort
+#' @returns a vector of dispersion parameters of length of the `delay_pmf` -1
+#' @export
+#'
+#' @examples
+#' triangle <- matrix(
+#'   c(
+#'     70, 40, 20, 5,
+#'     80, 50, 10, 10,
+#'     100, 40, 31, 20,
+#'     95, 45, 21, NA,
+#'     82, 42, NA, NA,
+#'     70, NA, NA, NA
+#'   ),
+#'   nrow = 6,
+#'   byrow = TRUE
+#' )
+#' delay_pmf <- get_delay_estimate(
+#'   triangle = triangle,
+#'   max_delay = 3,
+#'   n_history = 4
+#' )
+#' disp_params <- estimate_uncertainty(
+#'   triangle_to_nowcast = triangle,
+#'   delay_pmf = delay_pmf,
+#'   n_history_dispersion = 5
+#' )
+estimate_uncertainty <- function(triangle_to_nowcast,
+                                 delay_pmf,
+                                 n_history_dispersion = nrow(triangle_to_nowcast) - 1) { # nolint
+
+  n_horizons <- length(delay_pmf) - 1
+
+  # Add validation for n_history_dispersion, must be less than
+  # number of rows in the triangle minus one
+  .validate_triangle(triangle_to_nowcast)
+  .validate_delay_and_triangle(
+    triangle = triangle_to_nowcast,
+    delay_pmf = delay_pmf
+  )
+  if (n_history_dispersion > nrow(triangle_to_nowcast) - 1) {
+    cli_abort(
+      message = c(
+        "Triangle to nowcast does not contain sufficient rows to ",
+        "estimate uncertainty from `n_history_dispersion` observations. Either",
+        "pass in a triangle of more rows or lower the `n_history_dispersion`"
+      )
+    )
+  }
+
+  # Get the truncated matrix of observations you will use to estimate the
+  # dispersion (get rid of early rows that we're not using and add NAs to
+  # bottom right of the triangle)
+  matr_observed <- triangle_to_nowcast[
+    (nrow(triangle_to_nowcast) - n_history_dispersion):
+    nrow(triangle_to_nowcast),
+  ] |>
+    .replace_lower_right_with_NA()
+
+  results <- lapply(seq_len(n_history_dispersion), function(t) {
+    # Truncate the matrix observed by ignoring rows after t, replace rows that
```

Comment:
as before anything in a loop should be its own function so we can test it

Reply:
Will rewrite and add tests

Reply:
Did this, see https://github.com/epinowcast/baselinenowcast/blob/09fc7f241db416218759a23e69f167697103014c/R/estimate_uncertainty.R#L164 + tests